### ISSUE TITLE
Regression: Custom Predicates can't be passed lazily in a configure block

### ISSUE THREAD

**Issue by jodosha** [dry-rb/dry-validation#183]
After the recent changes on custom predicates: the following code doesn't work anymore:

```ruby
require 'bundler/setup'
require 'dry-validation'

module SharedPredicates
  include Dry::Logic::Predicates

  predicate(:accepted?) do |actual|
    actual && actual == true
  end
end

base = Dry::Validation.Schema(build: false) do
  required(:terms).filled(:bool?, :accepted?)
end

base.configure do |config|
  config.predicates = SharedPredicates
end

schema = Dry::Validation.Schema(build: false, rules: base.rules)
schema.new.call(terms: true)
```

Please note this is a simplified version of the code that `hanami-validations` uses internally. When `base` is initialized it checks immediately if the predicates used in the block are known or not. This raises an exception.

**Comment by jodosha** [comment]
I tried to simplify a lot with:

```ruby
require 'bundler/setup'
require 'dry-validation'

module SharedPredicates
  include Dry::Logic::Predicates

  predicate(:accepted?) do |actual|
    actual && actual == true
  end
end

schema = Dry::Validation.Schema(predicates: SharedPredicates) do
  required(:terms).filled(:bool?, :accepted?)
end

schema.call(terms: true)
```

But still doesn't work.

**Comment by jodosha** [comment]
My goal with `hanami-validations` is:

```ruby
class Signup
  include Hanami::Validations
  predicates SharedPredicates

  validations do
    required(:terms).filled(:bool?, :accepted?)
  end
end
```

Instead of the following syntax, which works but it's less tidy.

```ruby
class Signup
  include Hanami::Validations

  validations do
    configure do
      predicates SharedPredicates
    end

    required(:terms).filled(:bool?, :accepted?)
  end
end
```

**Comment by solnic** [comment]
Right, we validate if predicates are valid and they must be configured prior defining schema rules. I also added a new interface for setting custom predicate module: `configure { predicates(SharedPredicates) }`, but it must be called before defining rules (same applies to defining custom predicates methods in configure block, it must go before rules).

I'll update hanami-validations and send a PR, this shouldn't be a blocker for you. This change is a huge improvement as we have access to predicates when inside DSL and we can validate their names, arity and more in the near future. We can also cache built-in predicates so that we're not creating new rules instances for each schema (which will be a nice perf boost and mem-usage reduction).

**Comment by jodosha** [comment]
@solnic Thank you. I can update myself, as long at least this syntax will work: https://github.com/dry-rb/dry-validation/issues/183#issuecomment-226132316

**Comment by jodosha** [comment]
@solnic A gentle reminder to not release 0.8 if we haven't sorted this out. It's a blocking issue for Hanami. Thanks 💚 .

**Comment by solnic** [comment]
I'll figure it out prior release. Thanks for reminder :)

**Comment by solnic** [comment]
@jodosha I'm wondering now why do you set up a base schema by default and copy its rules to the other one? A simple fix would be to configure its predicates too

**Comment by jodosha** [comment]
@solnic Because I want to support useful defaults like: CSRF protection token param (`_csrf_param`), application wide predicates and I18n settings.

That code is too complex because of the inline predicates, that it's probably better to remove.

**Comment by solnic** [comment]
@jodosha I'll make this work better for you later this week, just need to better understand what exactly is needed. We should simplify the way your custom predicates are being setup. I'll figure something out :)
